### PR TITLE
hwdef for Skystars F405 DJI

### PR DESCRIPTION
**Pull request by labdegaragem**
This board isn't available yet.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+# MCU class and specific type.
+MCU STM32F4xx STM32F405xx
+
+HAL_CHIBIOS_ARCH_F405 1
+
+# board ID for firmware load
+# copied from the same board, must be changed in the future
+# to avoid conflicts
+APJ_BOARD_ID 1019
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat:14]: Please reserve a new  board ID now - and put it here.

What's the ETA on the board?

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat (hunk 1 above)
Comment (labdegaragem) [reply libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat]: Hello Peter!
Thanks for the support.
Reserved ID 146, next on [this list](https://github.com/ArduPilot/ardupilot/blob/master/Tools/AP_Bootloader/board_types.txt).
I'm new to this so I'm not sure how to include this number so bootloader can recognise it as a valid ID.
Please, let me know the next steps.
About the ETA, the board is already in the market. I just ported Ardupilot to it since it wasn't available yet.

Comment (labdegaragem) [reply libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat]: It's important to mention I couldn't make Smart Audio work yet (just in Betaflight with same connections and hardware). My partner asked for help [here](https://www.facebook.com/groups/ArduPilot.org/posts/4280995325256961/).

**Review by labdegaragem — COMMENTED** [review]
Changes as requested.
Please check additional comments I've made.
Thank you!

**Comment by davidbuzz** [comment]
What's the ETA on the board?

**Comment by labdegaragem** [comment]
> What's the ETA on the board?

Sorry. I'm new to this... what does ETA refer to?

**Comment by davidbuzz** [comment]
eta = Estimated Time of arrival. ie... when will this board be generally avail for purchase?

File: libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+# Based on Mamba F405 MK2 from jeanphilippehell
+# Pin information from Betaflight
+
+# MCU class and specific type.
+MCU STM32F4xx STM32F405xx
+
+HAL_CHIBIOS_ARCH_F405 1
+
+# board ID for firmware load
+APJ_BOARD_ID 146
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat:12]: This needs to be a unique number for this flight controller.  That means adding an entry to `Tools/AP_Bootloader/board_types.txt`, which you can do with a separate pull request - or another patch in here.

File: libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+define STORAGE_FLASH_PAGE 1
+# flash size
+FLASH_SIZE_KB 1024
+# reserve 32k for bootloader and 32k for flash storage
+FLASH_RESERVE_START_KB 64
+
+# To complementary channels work we define this
+#define STM32_PWM_USE_ADVANCED TRUE
+
+define BOARD_PWM_COUNT_DEFAULT 4
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat:146]: This can be removed

**Review by peterbarker — CHANGES_REQUESTED** [review]
I note that this is available for purchase, so it would be good to get this merged.

Can't merge it without a unique board ID, 'though!

File: libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat (hunk 2 above)
Comment (labdegaragem) [reply libraries/AP_HAL_ChibiOS/hwdef/SkystarsF405DJI/hwdef.dat:12]: Done!

**Comment by peterbarker** [comment]
@labdegaragem I've rebased and squash the commits.

Has this been tested on a board?

**Comment by labdegaragem** [comment]
Yes, it was tested in a quad I made with a friend.

**Review by tridge — CHANGES_REQUESTED** [review]
this seems to be missing the hwdef-bl.dat for the bootloader?

**Comment by tridge** [comment]
a README.md with the pinout and diagrams is also nice

**Comment by labdegaragem** [comment]
I used any available *_BL.HEX file and STM32Cube to load the bootloader to the STM32F405 on the board. Then, compiled and uploaded the above HWDEF file using WAF. Not sure if this is the best way to do it, but it worked fine.
Indeed the Readme file would be awesome. Not sure when I will have time for that though. Sorry.

**Comment by peterbarker** [comment]
Whoops!  This one seems to have been forgotten!

This seems to be missing a hardware definition file for the bootloader.

@labdegaragem are you still interested in pursuing this one?

**Comment by peterbarker** [comment]
No response, please re-open if you're keen, @labdegaragem
